coresimd/x86_64/mod.rs

```diff
@@ -38,3 +43,64 @@ pub use self::bswap::*;
 
 mod rdrand;
 pub use self::rdrand::*;
+
+/// Compare and exchange 16 bytes (128 bits) of data atomically.
+///
+/// This intrinsic corresponds to the `cmpxchg16b` instruction on x86_64
+/// processors. It performs an atomic compare-and-swap, updating the `ptr`
+/// memory location to `val` if the current value in memory equals `old`.
+///
+/// # Return value
+///
+/// This function returns the previous value at the memory location. If it is
+/// equal to `old` then the memory was updated to `new`.
+///
+/// # Memory Orderings
+///
+/// This atomic operations has the same semantics of memory orderings as
+/// `AtomicUsize::compare_exchange` does, only operating on 16 bytes of memory
+/// instead of just a pointer.
+///
+/// For more information on memory orderings here see the `compare_exchange`
+/// documentation for other `Atomic*` types in the standard library.
+///
+/// # Unsafety
+///
+/// This method is unsafe because it takes a raw pointer and will attempt to
+/// read and possibly write the memory at the pointer. The pointer must also be
+/// aligned on a 16-byte boundary.
```

Comment:
We probably want a `debug_assert!` here that checks that the pointer is 16-byte aligned. This won't be easily usable through `libcore`/`libstd`, but the `libstd` tests are run in CI with debug assertions enabled and this method might be used to implement some of the std atomic types, so it would at least be useful for that.